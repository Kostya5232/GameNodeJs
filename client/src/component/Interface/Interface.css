.back-arrow-interface {
    background-image: url("./image/back.png");
    background-size: contain;
    background-repeat: no-repeat;
    position: absolute;
    left: 0;
    top: 4px;
    width: 2vw;
    height: 2vw;
}
.settings-arrow-interface {
    background-image: url("./image/settings.png");
    background-size: contain;
    background-repeat: no-repeat;
    position: absolute;
    left: 2.4vw;
    top: 4px;
    width: 2vw;
    height: 2vw;
}
.player-info {
    position: absolute;
    top: 7vh;
    left: 1vw;
}

.player-icon-interfaceSportik,
.player-icon-interfaceHumanitarian,
.player-icon-interfaceTheechguy {
    position: relative;
    background-image: url("./image/playerIconSportik.png");
    background-size: contain;
    margin-top: 99.8%;
    background-repeat: no-repeat;
    width: 4vw;
    top: -4.3vw;
    height: 4vw;
}
.player-icon-interfaceHumanitarian {
    background-image: url("./image/playerIconHumanitarian.png");
}
.player-icon-interfaceTheechguy {
    background-image: url("./image/playerIconTheechguy.png");
}

.player-level-interface {
    background-color: #898888;
    position: absolute;
    left: 0;
    top: 0;
    border-radius: 50%;
    width: 1vw;
    height: 1vw;
    font-size: 0.8vw;
    text-align: center;
}
/* .player-hp-interface {
    background: linear-gradient(to right, red 50%, #898888 50%);
    position: absolute;
    right: -10vw;
    bottom: 0;
    color: white;
    width: 9vw;
    height: 1vw;
    font-size: 0.8vw;
} */

.map-info {
    position: absolute;
    right: 0;
    top: 0;
}
.miniMap-interface {
    border: 3px solid #000;
    background: #d9d9d9;
    width: 10vw;
    height: 10vw;
    text-align: center;
    display: table-cell;
    vertical-align: middle;
}
.game-info-interface {
    border: 3px solid #000;
    background: #d9d9d9;
    position: absolute;
    right: 10vw;
    top: 0;
    width: 10vw;
    height: 2.5vw;
    display: flex;
}
.coin-interface,
.time {
    font-size: 1.8vw;
    padding-left: 1vw;
}
.coin-icon-interface {
    background-image: url("./image/coin.png");
    background-size: contain;
    background-repeat: no-repeat;
    width: 1.5vw;
    height: 1.5vw;
    margin-top: 0.6vw;
    margin-left: 0.2vw;
}

.I {
    position: absolute;
    top: -5vw;
    font-size: 0.7vw;
    left: 107%;
    height: 15%;
    width: 300%;
    /* background: linear-gradient(to right, red var(--elem-hp, 100%), #898888 var(--elem-hp, 100%)); */
}
.HP {
    color: white;
    margin-top: 61%;
    clip-path: polygon(0 0, 100% 0, 89% 100%, 0 100%);
}
.BossXP {
    position: absolute;
    font-size: 1.7vw;
    width: 200px;
    height: 200px;
    top: 1vw;
    right: 50%;
    color: red;
}
