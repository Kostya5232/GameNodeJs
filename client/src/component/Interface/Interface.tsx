import React, { useContext, useEffect, useRef, useState } from "react";
import { MediatorContext, ServerContext } from "../../App";
import Chat from "./component/Chat/Chat";
// import TaskSelection from "./component/TaskSelection/TaskSelection";
import { TGamer, TMobs, TScene } from "../../modules/Server/types";
import "./Interface.css";
import ParametersGame from "./component/ParametersGame/ParametersGame";
import TaskSelection from "./component/TaskSelection/TaskSelection";

const Interface: React.FC = () => {
    const server = useContext(ServerContext);
    const mediator = useContext(MediatorContext);
    const [infoMobs, setInfoMobs] = useState<TMobs[] | null>(null);
    const timeRef = useRef<HTMLDivElement>(null);
    const [gamers, setGamers] = useState<any>(null);
    const [questionFlag, setQuestionFlag] = useState<boolean>(true);
    const [timer, setTimer] = useState({ seconds: 0, minutes: 0 });
    const parametersGameRef = useRef<HTMLDivElement>(null);
    const [showParametersGame, setShowParametersGame] = useState(false);

    const handleSettingsClick = () => {
        setShowParametersGame(true);
    };
    const handleParametersGameClick = () => {
        setShowParametersGame(false);
    };
    const handleOutsideClick = (event: MouseEvent) => {
        if (parametersGameRef.current && !parametersGameRef.current.contains(event.target as Node)) {
            setShowParametersGame(false);
        }
    };
    useEffect(() => {
        document.addEventListener("mousedown", handleOutsideClick);
        const updateTime = () => {
            setTimer((prevTimer) => {
                let seconds = prevTimer.seconds + 1;
                let minutes = prevTimer.minutes;

                if (seconds === 60) {
                    minutes++;
                    seconds = 0;
                }

                return { seconds, minutes };
            });
        };

        let timeInterval = setInterval(updateTime, 1000);
        return () => {
            document.removeEventListener("mousedown", handleOutsideClick);
            clearInterval(timeInterval);
        };
    }, []);

    useEffect(() => {
        const { GET_GAMERS } = mediator.getEventTypes();
        const { GET_MOBS } = mediator.getEventTypes();

        const getGamersHandler = (data: any) => {
            setGamers(data);
        };
        const getMobsHandler = (data: any) => {
            setInfoMobs(data);
        };

        mediator.subscribe(GET_GAMERS, getGamersHandler);
        mediator.subscribe(GET_MOBS, getMobsHandler);

        return () => {
            mediator.unsubscribe(GET_GAMERS, getGamersHandler);
            mediator.unsubscribe(GET_MOBS, getMobsHandler);
        };
    });

    // вызов окошка задания
    if (!questionFlag && mediator.triger) {
        if (mediator.tim < 13) {
            mediator.tim += 1;
        } else {
            setQuestionFlag(true);
            mediator.tim = 0;
        }
    }

    return (
        <div className="Interface-container">
            <div className="back-arrow-interface"></div>
            <div onClick={handleSettingsClick} className="settings-arrow-interface"></div>
            {showParametersGame && (
                <div ref={parametersGameRef}>
                    <ParametersGame />
                </div>
            )}

            <div className="player-info">
                {gamers && gamers[0].person_id === 0 ? (
                    <div className="player-icon-interfaceSportik"></div>
                ) : gamers && gamers[0].person_id === 2 ? (
                    <div className="player-icon-interfaceHumanitarian"></div>
                ) : gamers && gamers[0].person_id === 1 ? (
                    <div className="player-icon-interfaceTheechguy"></div>
                ) : (
                    <></>
                )}

                <div className="I">
                    {gamers ? (
                        gamers.map((elem: any, index: number) => (
                            <div
                                key={index}
                                className="HP"
                                style={{
                                    background: `linear-gradient(to right, red ${elem.hp}%, #898888 ${elem.hp}%)`,
                                }}
                            >
                                {elem.name} : {elem.hp}/100HP
                            </div>
                        ))
                    ) : (
                        <></>
                    )}
                </div>

                <div className="player-level-interface">1</div>
            </div>

            <div className="map-info">
                <div className="miniMap-interface">Мини карта</div>

                <div className="game-info-interface">
                    <div className="coin-interface">0</div>
                    <div className="coin-icon-interface"></div>
                    <div ref={timeRef} className="time">
                        {timer.minutes < 10 ? "0" + timer.minutes : timer.minutes} :{" "}
                        {timer.seconds < 10 ? "0" + timer.seconds : timer.seconds}
                    </div>
                </div>
            </div>
            {mediator.triger && questionFlag ? <TaskSelection setQuestionFlag={setQuestionFlag} /> : <></>}
            {mediator.triger ? (
                <div className="BossXP">{infoMobs ? <div>BossXP: {infoMobs[0].hp}</div> : <></>}</div>
            ) : (
                <></>
            )}

            <Chat />
        </div>
    );
};

export default Interface;
