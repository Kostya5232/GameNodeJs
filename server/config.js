const CONFIG = {
    // dev
    DATABASE: {
        NAME: "studfront",
        HOST: "localhost",
        PORT: 5432,
        USER: "postgres",
        PASS: "111",
    },

    MESSAGES: {
        SEND_MESSAGE: "SEND_MESSAGE",
        GET_MESSAGES: "GET_MESSAGES",
    },
    MEDIATOR: {
        EVENTS: {
            TEST: "TEST",
            // SERVER_ERROR: "SERVER_ERROR",
        },
        TRIGGERS: {},
    },
};

module.exports = CONFIG;
