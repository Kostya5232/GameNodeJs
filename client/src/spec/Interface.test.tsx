import { render, screen } from "@testing-library/react"; 
import Interface from "../component/Interface/Interface"; 
import { MediatorContext, ServerContext } from "../App"; 
import { Mediator, Server } from "../modules"; 
import { HOST, MEDIATOR } from "../config"; 
 
const mediator = new Mediator(MEDIATOR); 
const server = new Server(HOST, mediator); 
jest.useFakeTimers(); 
 
test("Interface snapshot", () => { 
 
  const { asFragment } = render( 
    <MediatorContext.Provider value={mediator}> 
      <ServerContext.Provider value={server}> 
        <Interface /> 
      </ServerContext.Provider> 
    </MediatorContext.Provider> 
  ); 

    expect(asFragment()).toMatchSnapshot(); 
 
  jest.advanceTimersByTime(1000);   
  expect(asFragment()).toMatchSnapshot(); 
 
  jest.useRealTimers(); 
});