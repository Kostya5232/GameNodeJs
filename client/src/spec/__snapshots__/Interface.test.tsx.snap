// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interface snapshot 1`] = `
<DocumentFragment>
  <div
    class="Interface-container"
  >
    <div
      class="back-arrow-interface"
    />
    <div
      class="settings-arrow-interface"
    />
    <div
      class="player-info"
    >
      <div
        class="I"
      />
      <div
        class="player-level-interface"
      >
        1
      </div>
    </div>
    <div
      class="map-info"
    >
      <div
        class="miniMap-interface"
      >
        Мини карта
      </div>
      <div
        class="game-info-interface"
      >
        <div
          class="coin-interface"
        >
          0
        </div>
        <div
          class="coin-icon-interface"
        />
        <div
          class="time"
        >
          00 : 00
        </div>
      </div>
    </div>
    <div
      class="Chat-container"
    >
      <div
        class="chat-interface"
      >
        Чат
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Interface snapshot 2`] = `
<DocumentFragment>
  <div
    class="Interface-container"
  >
    <div
      class="back-arrow-interface"
    />
    <div
      class="settings-arrow-interface"
    />
    <div
      class="player-info"
    >
      <div
        class="I"
      />
      <div
        class="player-level-interface"
      >
        1
      </div>
    </div>
    <div
      class="map-info"
    >
      <div
        class="miniMap-interface"
      >
        Мини карта
      </div>
      <div
        class="game-info-interface"
      >
        <div
          class="coin-interface"
        >
          0
        </div>
        <div
          class="coin-icon-interface"
        />
        <div
          class="time"
        >
          00 : 01
        </div>
      </div>
    </div>
    <div
      class="Chat-container"
    >
      <div
        class="chat-interface"
      >
        Чат
      </div>
    </div>
  </div>
</DocumentFragment>
`;
